src/transformers/image_utils.py

```diff
@@ -62,6 +62,121 @@
+
+    See: https://numpy.org/doc/stable/user/basics.dispatch.html
+
+    Args:
+        data (`ImageInput`):
+            The image data.
+        data_format (`ChannelDimension`, *optional*):
+            The channel dimension format of the image. If `None`, will be inferred from the image data.
+        num_channels (`int`, *optional*):
+            The number of channels in the image. If `None`, will be inferred from the image data.
+    """
+
+    def __init__(
+        self, data: ImageInput, data_format: Optional[ChannelDimension] = None, num_channels: int = None
+    ) -> None:
+        if isinstance(data, PIL.Image.Image):
+            if data.mode in ("1", "L", "P", "I;16", "I", "F"):
+                data_format = ChannelDimension.NONE
+            else:
+                # PIL images are always channels last when converted to numpy arrays
+                data_format = ChannelDimension.LAST
+
+        data = np.array(data)
+
+        if data_format is None:
+            data_format = infer_channel_dimension_format(data)
+
+        if data_format == ChannelDimension.NONE:
+            num_channels = 0
+            height = data.shape[-2]
+            width = data.shape[-1]
+        elif data_format == ChannelDimension.FIRST:
+            num_channels = data.shape[-3]
+            height = data.shape[-2]
+            width = data.shape[-1]
+        elif data_format == ChannelDimension.LAST:
+            num_channels = data.shape[-1]
+            height = data.shape[-3]
+            width = data.shape[-2]
+        else:
+            raise ValueError("Invalid channel dimension format")
+
+        self._data = data
+        self._data_format = data_format
+        self._num_channels = num_channels
+        self._height = height
+        self._width = width
+
+    def __getattribute__(self, __name: str) -> Any:
+        if __name in ("_data",):
+            return super().__getattribute__(__name)
+
+        if hasattr(self._data, __name):
+            result = getattr(self._data, __name)
+        else:
+            result = super().__getattribute__(__name)
+        return _output_wrapper(result)
+
+    def __getitem__(self, key: Any) -> Any:
+        result = self._data[key]
+        return _output_wrapper(result)
+
+    def __setitem__(self, key: Any, value: Any) -> None:
+        self._data[key] = value
+
+    def __array__(self, dtype=None):
+        return self._data.astype(dtype)
+
+    def __array_ufunc__(self, ufunc, method, *inputs: Iterable[Any], **kwargs: Mapping[str, Any]) -> Any:
+        if not all(isinstance(input, (np.ndarray, ImageObject) + np.ScalarType) for input in inputs):
+            return NotImplemented
+
+        scalars = ()
+        for input in inputs:
+            if isinstance(input, ImageObject):
+                scalars += (input._data,)
+            else:
+                scalars += (input,)
+        result = getattr(ufunc, method)(*scalars, **kwargs)
+        return _output_wrapper(result)
```

Comment:
Do we need to consider both `np.ndarray` and `ImageObject` here?

Reply:
For the input or the output or something else? 

The input we get when collecting the `scalars` and the output we convert to `ImageArray` if the output is `np.ndarray`